---
title: "MAMBO Tutorial"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is a tutorial for running metabarcoding data sets through MAMBO (Metabarcoding Analysis of Modeled Bayesian Occurrences). You should start with two ASV tables, each of which has columns as samples and rows as ASVs.

```{r libraries}
library(mambo)
```

## ASV Tables

ASV count tables should be loaded as R data frames or matrices. The values in the tables represent the occurrence (number of reads) fore each ASV (rows) in each sample (columns). The table should have both row and column names uniquely identifying ASVs and samples.

We supply two example ASV count tables in **mambo**: '**counts.16s**' and '**counts.18s**'.
```{r data}
counts.16s[1:8, 1:4]
```

## Run MAMBO 

To run the main function, **mambo()**, we need to supply labels for the response and predictor datasets and the response and predictor count tables, along with the number of replicates (**nrep**) and other MCMC parameters (see **?mambo**).  
```{r mambo, include=FALSE}
result <- mambo(
  resp.label = '18s', 
  resp.counts = counts.18s, 
  pred.label = '16s', 
  pred.counts = counts.16s, 
  nrep = 5, 
  chains = 10
)
```

## Examine Results

The return value of the **mambo()** function is list, containing five elements:   

1) The filename of the .RDS file that was written that contains this **mambo()** output list:  
```{r rds}
result$filename
```

2) The labels that were provided to the function for each locus:
```{r}
result$labels
```

3) The MCMC parameters:
```{r}
result$params
```

4) The run times for the model:
```{r}
result$run.time
```

5) A list of the results for every replicate.  Each element in this list contains the PCA results for both loci ($pca)
```{r}
str(result$reps[[1]]$pca[[1]])
```

...and a matrix with summaries of the posteriors of the Bayesian linear model:
```{r}
str(result$reps[[1]]$post.smry)
```

...and a list of the actual posterior sample from the Bayesian linear model:
```{r}
str(result$reps[[1]]$post.list)
```

## Plot the PCAs for each marker gene ("locus")
```{r PCA_16S}
plotPCs(result, '16s')
```

```{r PCA_18S}
plotPCs(result, '18s')
```

## Check which ASVs are outliers, ie driving the separations
```{r outliers_16S}
outliers.16s.pc1 <- outlierLoadings(result, locus = '16s', pc = 1)
outliers.18s.pc1 <- outlierLoadings(result, locus = '18s', pc = 1)

head(outliers.16s.pc1)
```

```{r outliers_18S}
head(outliers.18s.pc1)
```

### Identify taxonomy of outlier ASVs
```{r taxa}
# Add column name to match ASV tibble column name if it doesn't already
taxa.16s$asv <- rownames(taxa.16s)
taxa.18s$asv <- rownames(taxa.18s)

# Merge the tibbles
asvs_out_16s_tax <- dplyr::left_join(outliers.16s.pc1, taxa.16s, by = "asv")
asvs_out_18s_tax <- dplyr::left_join(outliers.18s.pc1, taxa.18s, by = "asv")

head(asvs_out_16s_tax)
```

